janus_core/src/retries.rs

```diff
@@ -0,0 +1,315 @@
+//! Provides a simple interface for retrying fallible HTTP requests.
+
+use backoff::{future::retry, ExponentialBackoff};
+use futures::Future;
+use reqwest::StatusCode;
+use std::time::Duration;
+use tracing::{debug, warn};
+
+/// Traverse chain of source errors looking for an `std::io::Error`.
+fn find_io_error(original_error: &dyn std::error::Error) -> Option<&std::io::Error> {
+    let mut cause = original_error.source();
```

Comment:
nit: if `original_error` is already a `std::io::Error`, this function won't return it. This is only a nit because I think that doesn't matter for how the function is currently used. The easiest fix is probably to change this line to something like `let mut cause = Some(original_error)`, i.e. start walking the chain from `original_error` rather than `original_error`'s source.

Reply:
Generalizing as you suggest is surprisingly tricky because of the `'static` lifetime in the signature of `std::error::Error::source`. In this module, we know that `original_error` will only ever be `reqwest:Error`, so I think the easiest thing is to change the signature of `find_io_error` to reflect that.